infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -158,15 +165,19 @@ protected void processRequest(RequestMsg req, ChannelHandlerContext ctx, IServer
     /**
      * Service an incoming request for maximum global address the log unit server has written.
      */
-    @ServerHandler(type = CorfuMsgType.TAIL_REQUEST)
-    public void handleTailRequest(CorfuPayloadMsg<TailsRequest> msg, ChannelHandlerContext ctx, IServerRouter r) {
-        log.debug("handleTailRequest: received a tail request {}", msg);
-        batchWriter.<TailsResponse>addTask(TAILS_QUERY, msg)
-                .thenAccept(tailsResp -> {
-                    r.sendResponse(ctx, msg, CorfuMsgType.TAIL_RESPONSE.payloadMsg(tailsResp));
-                })
+    @RequestHandler(type = PayloadCase.TAIL_REQUEST)
+    public void handleTailRequest(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        log.debug("handleTailRequest[{}]: received a tail request {}",
+                req.getHeader().getRequestId(), TextFormat.shortDebugString(req));
+
+        batchWriter.<TailsResponse>addTask(BatchWriterOperation.Type.TAILS_QUERY, req)
+                .thenAccept(tailsResp ->
+                    // Note: we reuse the request header as the ignore_cluster_id and
+                    // ignore_epoch fields are the same in both cases.
```

Comment:
This is incorrect. You can get receive a message on epoch 1, then by the time you reply the server is on epoch 2. The client should be able to detect that. The message can be queued internally while the epoch of the router changes.